experimental/Schemes/SpaceGerms.jl

```diff
@@ -283,21 +981,275 @@ function Base.union(X::AbsSpaceGerm, Y::AbsSpaceGerm)
+
+Returns the space germ (Y,p) for a given germ (X,p) and the closed embedding of (Y,p) into (X,p), where Y is the singular locus of X.
+
+# Examples:
+```jldoctest
+julia> X = affine_space(QQ,3);
+
+julia> R = coordinate_ring(X);
+
+julia> (x,y,z) = gens(R);
+
+julia> Q,_ = quo(R,ideal(R,[x^2+y^2]));
+
+julia> Y = SpaceGerm(Spec(Q),[0,0,0]);
+
+julia> XSL, incSL = singular_locus(Y);
+
+julia> XSL
+Spectrum
+  of localization
+    of quotient
+      of multivariate polynomial ring in 3 variables x1, x2, x3
+        over rational field
+      by ideal(x1^2 + x2^2, x2, x1)
+    at complement of maximal ideal of point (0, 0, 0)
+
+julia> incSL
+Morphism
+  from [x1, x2, x3]  Spec of localization of quotient of multivariate polynomial ring at complement of maximal ideal
+  to   [x1, x2, x3]  Spec of localization of quotient of multivariate polynomial ring at complement of maximal ideal
+given by the pullback function
+  x1 -> x1
+  x2 -> x2
+  x3 -> x3
+
+```
+"""
 function singular_locus(X::AbsSpaceGerm)
   S, inc = singular_locus(underlying_scheme(X))
   Sgerm = SpaceGerm(S)
   return Sgerm, ClosedEmbedding(SpecMor(Sgerm, X, pullback(inc), check=false), image_ideal(inc), check=false)
 end
 
-function subscheme(X::SpaceGerm, I::Ideal)
+## note: subgerms of hypersurface and complete intersection germs are simply space germs
+@doc raw"""
+    subgerm(X::AbsSpaceGerm, I::Ideal) --> SpaceGerm
+
+Returns the space germ (Y,p) of (X,p) defined by the ideal I in the local ring of X at p
+
+!!! note: (Y,p) is of type SpaceGerm, even if (X,p) is a HypersurfaceGerm or a CompleteIntersectionGerm.
+"""
+function subgerm(X::AbsSpaceGerm, I::Ideal)
   base_ring(I) === OO(X) || error("ideal does not belong to the correct ring")
   Y = subscheme(underlying_scheme(X), I)
   return SpaceGerm(Y)
 end
 
+@doc raw"""
+    is_isolated_singularity(X::AbsSpaceGerm)
+
+Returns whether `(X,p)` has at most an isolated singularity.
+```jldoctest
+julia> X = affine_space(QQ,3);
+
+julia> R = coordinate_ring(X);
+
+julia> (x,y,z) = gens(R);
+
+julia> Q,_ = quo(R,ideal(R,[x^2+y^2]));
+
+julia> Y = SpaceGerm(Spec(Q),[0,0,0]);
+
+julia> is_isolated_singularity(Y)
+false
+
+```
+"""
+@attr Bool function is_isolated_singularity(X::AbsSpaceGerm)
+  dim(singular_locus(X)[1]) < 1 || return false
+  return true
+end
+
+##############################################################################
+# milnor_number, milnor_algebra for IHS
+# milnor_number for ICIS
+# -- beyond this we do no longer have a bouquet of spheres of same dimension
+##############################################################################
+@doc raw"""
+    milnor_algebra(X::HypersurfaceGerm)
+
+Returns the local Milnor algebra of `(X,p)` at p
+"""
+function milnor_algebra(X::HypersurfaceGerm)
+  R = localized_ring(OO(X))
+  ## milnor number independent of choice of representative
+  ## hence choose a polynomial representative for easier computation
+  f_poly = numerator(X.f)
```

Comment:
I suggest to use `defining_ring_elem(X)` instead of `X.f`. This makes the code more easily maintainable in the long run. In general, only the getters themselves should be allowed to access the internals of a struct.